tensorflow/compiler/tf2tensorrt/convert/timing_cache.cc

```diff
@@ -0,0 +1,81 @@
+/* Copyright 2022 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+#if GOOGLE_CUDA && GOOGLE_TENSORRT
+
+#include "tensorflow/compiler/tf2tensorrt/convert/timing_cache.h"
+
+#include <unordered_map>
+
+#include "tensorflow/compiler/tf2tensorrt/common/utils.h"
+#include "tensorflow/core/platform/errors.h"
+#include "third_party/tensorrt/NvInfer.h"
+
+namespace tensorflow {
+namespace tensorrt {
+namespace convert {
+
+StatusOr<TimingCacheRegistry::TimingCachePtr> TimingCacheRegistry::LookUp(
+    const string& name, nvinfer1::IBuilderConfig* builder_config) {
+  TRT_ENSURE(builder_config != nullptr);
+  mutex_lock scoped_lock(mu_);
+  if (map_.find(name) != map_.end()) {
+    const std::vector<uint8_t>& data = map_[name];
+    return std::unique_ptr<nvinfer1::ITimingCache>(
+        builder_config->createTimingCache(data.data(), data.size()));
+  }
+
+  // If no such timing cache exists, create a new timing cache.
+  return std::unique_ptr<nvinfer1::ITimingCache>(
+      builder_config->createTimingCache(nullptr, 0));
+}
+
+void TimingCacheRegistry::Upsert(const string& name,
+                                 nvinfer1::ITimingCache* cache) {
+  nvinfer1::IHostMemory* memory = cache->serialize();
+  if (memory == nullptr) {
+    return;
+  }
+
+  if (map_.find(name) == map_.end()) {
+    // If the timing cache with the given name does not exist, emplace the
+    // serialized buffer.
+    std::vector<uint8_t> mem(memory->size());
+    std::copy_n(static_cast<uint8_t*>(memory->data()), memory->size(),
+                mem.begin());
+    {
+      mutex_lock scoped_lock(mu_);
+      map_.emplace(name, std::move(mem));
+    }
+  } else {
+    // If the timing cache does exist, use the existing buffer.
+    mutex_lock scoped_lock(mu_);
+    std::vector<uint8_t>& mem = map_[name];
+    mem.resize(memory->size());
+    std::copy_n(static_cast<uint8_t*>(memory->data()), memory->size(),
+                mem.begin());
+  }
+  memory->destroy();
+}
+
+TimingCacheRegistry* GetTimingCacheRegistry() {
+  static TimingCacheRegistry* registry = new TimingCacheRegistry();
```

Comment:
I don't see how this memory is release. Shall we use a unique pointer here?

Reply:
I though this is what we were doing for all the other registry examples I looked at, but I will double check, thanks

Reply:
This will cause memory leak, right?

Reply:
Right... maybe I'm not understanding something about the pattern I'm observing. For example, see here: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/framework/op.cc#L260.

Reply:
There seem to be may examples, just search for the pattern `/= new.*;/` in all files `*registry*.cc`.

Reply:
So I gather the use of `new` without `delete` for global singletons  in the codebase is due to the [static initialization/deinitialization problem](https://isocpp.org/wiki/faq/ctors). In general, if we can guarantee that this global singleton will never be accessed from any other global object's destructor, then we should just be able to replace the 
```
static Registry *reg = new Registry();
return reg;
```
with
```
static Registry reg;
return &reg;
```

Looking elsewhere in the codebase by grepping `/= new.*;/` in all files `*registry*.cc`, it seems that the use of the first pattern to `new` global singletons and ignore the leaks (because the OS will reclaim the memory) is pretty prevalent pattern in Tensorflow. I guess it's possible there is some global statically initialized object that could potentially call arbitrary other global objects during construction? That would explain the pattern, and in which case we should follow it as well.

Reply:
You also mentioned `asan` issues. Potentially there is an ignore list for registries? Maybe if I just rename the file `timing_cache_registry.cc` if that will activate the ignore?

Reply:
According to google internal style guide [here](go/cstyle#common_patterns), we should use gtl::NoDestructor<TimingCacheRegistry>, but I couldn't find this template class in the TF tree. This may be an explanation that TF is using the pattern (static pointer, new), because it has a one-to-one trivial mapping to TimingCacheRegistry. I am fine with the current code here for this reason.